src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -626,7 +632,60 @@ private void MakeExplicitParameterTypeInferences(BoundExpression argument, TypeW
             }
         }
 
-        private bool MakeExplicitParameterTypeInferences(BoundTupleLiteral argument, TypeWithAnnotations target, ExactOrBoundsKind kind, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        private void MakeCollectionLiteralTypeInferences(
+            Binder binder,
+            BoundUnconvertedCollectionLiteralExpression argument,
+            TypeWithAnnotations target,
+            ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            if (target.Type is null)
+            {
+                return;
+            }
+
+            if (argument.Elements.Length == 0)
+            {
+                return;
+            }
+
+            if (!tryGetCollectionIterationType(binder, (ExpressionSyntax)argument.Syntax, target.Type, out TypeWithAnnotations targetElementType))
+            {
+                return;
+            }
+
+            foreach (var element in argument.Elements)
+            {
+                if (element.Kind == BoundKind.UnconvertedCollectionLiteralExpression)
+                {
+                    MakeCollectionLiteralTypeInferences(binder, (BoundUnconvertedCollectionLiteralExpression)element, targetElementType, ref useSiteInfo);
+                }
+                else
+                {
+                    var elementType = _extensions.GetTypeWithAnnotations(element);
+                    if (IsReallyAType(elementType.Type))
+                    {
+                        LowerBoundInference(elementType, targetElementType, ref useSiteInfo);
+                    }
+                }
+            }
+
+            static bool tryGetCollectionIterationType(Binder binder, ExpressionSyntax syntax, TypeSymbol collectionType, out TypeWithAnnotations iterationType)
+            {
+                var builder = new ForEachEnumeratorInfo.Builder();
+                BoundExpression collectionExpr = new BoundValuePlaceholder(syntax, collectionType);
+                return binder.GetEnumeratorInfoAndInferCollectionElementType(
```

Comment:
This makes me thing that we should do this up front, when we do the initial unconverted collection literal info calculation, and not here. If we're in some generic overload scenario where we need to perform a number of type inferences, this is going to become extremely expensive extremely quickly. Needing pass a binder into the MethodTypeInferrer is, to me, a signal that the bound node is missing a critical piece of information.

Reply:
The collection type in this case is the parameter type rather than the collection literal argument type, so we don't have a bound node in the tree that is directly associated with the type.

Reply:
How do we solve this for tuple literals? Presumably they have a similar problem when you have a signature that takes a `(T, T)`, but they were able to solve it without a binder.

Reply:
For the collection case, where we're inferring from a collection literal passed as an argument to a parameter that is a collection type, we need to determine the iteration type, and that requires a binder in particular for the foreach-able pattern cases. By doing that here, in `MethodTypeInferrer`, we essentially calculating the iteration type lazily, when we know we need it.